Add Multi Channel Transport Model (MCT)

schmoelder opened the pull request:
This PR implements a modified 2D-GRM to model plant root systems as described [here](http://dx.doi.org/10.1016/j.jtbi.2013.09.023)

## To-do
- [x] Define Interface (@jazib-hassan-juelich)
- [x] Implementation
    - [x] Take 2D-General Rate Model as template (rename and register classes)
    - [x] Adapt sparse matrix structure
https://github.com/modsim/CADET/blob/8fe0d0203a94242a0c51032704ab3585b14a1caf/src/libcadet/model/parts/TractorConvectionDispersionOperator.cpp#L1122
    - [x] Add fluxes in residual
- [x] Documentation (@HannahLanzrath, @lieres)
    - [x] Model
      - [x] Text + Equationss
      - [x] Update figure
    - [x] Interface
- [x] Add unittest
  - [x] add (dense) AD to enable AD Jacobian tests
- [x] Check if other changes in 2D-GRM have been made since creation of this branch (e.g. section dependent velocities)
- [x] Update exchange areas (for non-concentric channels)
- [x] Fix build for Ubuntu and MacOS

sleweke-bayer commented:
Where does the name `Tractor` come from? Maybe a more neutral name that can be understood by non-insiders would be fitting?
Suggestion: `CompartmentTransportModel`

Immudzen commented:
I agree with giving it a descriptive name. That also makes it easier to use the name or if it appears in a presentation, paper etc.

schmoelder commented:
Oh yes, it's just a working title because @hannahlanzrath is still checking with her supervisor what to call it.

schmoelder reviewed (commented):
Do we need  separate function for it or can we just read the parameter like any other model parameter?

jazib-hassan-juelich commented:
> I don't think we will have the decay terms here. It's easier to use the existing reactions interface.

agreed, but I just added the general framework and as soon as I have information from Hannah about the actual interface then I will made modifications accordingly.

jazib-hassan-juelich commented:
> Do we need separate function for it or can we just read the parameter like any other model parameter?

In my view, adding a specific configuration function to read the parameters of this particular model makes the tracibility of the current implementation a bit straight forward. Just my opinion. Else we can read them like the other parameters.

sleweke-bayer reviewed (commented):
That will work as a first shot but is going to be inefficient. 
I'd prefer setting the sparsity pattern based on the actual exchange matrix structure (i.e., only add non-zeros in the pattern for compartments that actually communicate).
To this end, `setSparsityPattern()` can be called in `configure()` instead of `configureModelDiscretization()`.

schmoelder commented:
> I'd prefer setting the sparsity pattern based on the actual exchange matrix structure (i.e., only add non-zeros in the pattern for compartments that actually communicate).

We fixed that by only adding the pattern only if the entry in the matrix is non-zero.

>  To this end, `setSparsityPattern()` can be called in `configure()` instead of `configureModelDiscretization()`.

@jazib-hassan-juelich @jayghoshter you had at the order of calling the various configure functions. Any comments on that?

Other than that, we're still awaiting tests from @hannahlanzrath but I hope it works now. 

We also implemented the exchange matrix s.t. it also would work for multiple components.

assuming e_{i,j,k} describes the transport of component k from compartment i to j, the matrix has the following form (assuming a 2 component system):
[
0,             0,              e_{0,1,0}, e_{0,1,1}, e_{0,2,0}, e_{0,2,1}, ...
e_{1,0,0}, e_{1,0,1}, 0,              0,             e_{1,2,0}, e_{1,2,1}, ...
...
]

@Immudzen also suggested we could take a similar approach to how we setup the connections, meaning that we only specify [compartment_from, compartment_to, component_index, exchange_rate].
This would simplify the user interface but I suggest we leave that for the python side to figure out.

schmoelder commented:
> > To this end, `setSparsityPattern()` can be called in `configure()` instead of `configureModelDiscretization()`.
> 
> @jazib-hassan-juelich @jayghoshter yesterday, you had a look at the order of calling the various configure functions. Any comments on that?

ping @jayghoshter @jazib-hassan-juelich

jazib-hassan-juelich commented:
> > > To this end, `setSparsityPattern()` can be called in `configure()` instead of `configureModelDiscretization()`.
> > 
> > 
> > @jazib-hassan-juelich @jayghoshter yesterday, you had a look at the order of calling the various configure functions. Any comments on that?
> 
> ping @jayghoshter @jazib-hassan-juelich
If I remember correctly, we first tried to implement the exchange matrix in the configure function, but then we faced the problem in the residualImpl [Function](https://github.com/modsim/CADET/blob/74059fd73b992acf5be12e22b6a57357ce4c0eb8/src/libcadet/model/parts/TractorConvectionDispersionOperator.cpp#L989 ) in [Line](https://github.com/modsim/CADET/blob/74059fd73b992acf5be12e22b6a57357ce4c0eb8/src/libcadet/model/parts/TractorConvectionDispersionOperator.cpp#L1144) that the exchange matrix had no entries which caused the simulation to fail. Due to this issue we defined the exchange matrix configureModelDiscretization() so that when this function is called exhange matrix is already defined.
Alongwith @Immudzen  and @jayghoshter  I looked deep into this issue and it relates to the order of calling these functions and when we try to change the order then broke the code. @jayghoshter  can add more to this.

jayghoshter commented:
Not much more to comment. As @jazib-hassan-juelich wrote, `configureModelDiscretization()` is called before `configure()` in `ModelBuilderImpl.cpp:183`. 

```
if (!model->configureModelDiscretization(paramProvider, *this) || !model->configure(paramProvider))
```

I think the issue was that `setSparsityPattern()` is called in `configureModelDiscretization` and needs `_exchangeMatrix` to be defined.

sleweke commented on src/libcadet/model/parts/TractorConvectionDispersionOperator.cpp:
```diff
@@ -0,0 +1,1526 @@
+
+					for (unsigned int comp = 0; comp < _nComp; ++comp)
+					{
+						const ParamType exchange_orig_dest_comp = static_cast<ParamType>(_exchangeMatrix[rad_orig * _nRad * _nComp + rad_dest * _nComp + comp]);
```
This will always be `active`. In fact, it will crash or produce garbage if `ParamType` is anything other than `active`. So no need for this function to be a template.

sleweke commented:
> I think the issue was that `setSparsityPattern()` is called in `configureModelDiscretization` and needs `_exchangeMatrix` to be defined.

Well, then don't call it at this point 😛 It makes more sense to read the exchange matrix in `configure()`. Instead of calling `setSparsityPattern()` in `configureModelDiscretization()`, just call it in `configure()` when everything is known.

sleweke commented on src/libcadet/model/parts/TractorConvectionDispersionOperator.cpp:
```diff
@@ -0,0 +1,1526 @@
+						// TODO: @Sam, please check this
+						if (wantJac)
+						{
+							_jacC.centered(offsetCur_orig, 0) += static_cast<double>(exchange_orig_dest_comp);
+							_jacC.centered(offsetCur_orig, offsetCur_dest) -= static_cast<double>(exchange_orig_dest_comp);
+							_jacC.centered(offsetCur_dest, 0) -= static_cast<double>(exchange_orig_dest_comp);
+							_jacC.centered(offsetCur_dest, offsetCur_orig) += static_cast<double>(exchange_orig_dest_comp);
```
We have changes in two different residuals (is this correct? Could already be accounted for in `_exchangeMatrix`). In each change, only one state vector item is involved. Hence, we should see two changes in the Jacobian:
* `d resCur_orig / d yCur_orig = exchange_orig_dest_comp`
* `d resCur_dest / d yCur_orig = -exchange_orig_dest_comp`

We also need to take into account that the second argument of `centered()` refers to the offset on the main diagonal of the matrix. That is, `centered(x,y)` refers to matrix item `mat_{x, x+y}`.

This leads to the code
```cpp
_jacC.centered(offsetCur_orig, 0) += static_cast<double>(exchange_orig_dest_comp);
_jacC.centered(offsetCur_dest, static_cast<int>(offsetCur_orig) - static_cast<int>(offsetCur_dest)) -= static_cast<double>(exchange_orig_dest_comp);
```
Since `resCur_orig` corresponds to `yCur_orig`, we get diagonal `0`. For the second, we need the offset from `resCur_dest` to `yCur_orig`. From our current position (diagonal `0`, equivalent to `resCur_dest`), we go back to the origin (`-offsetCur_dest`) and from there to `yCur_orig` (`+offsetCur_orig`). Note that we need to convert to `int` because `unsigned int` does not allow for negative offsets.

sleweke commented:
General remark: Please use consistent indentation (tabs?).

schmoelder commented:
sorry, I made a mess while trying to rebase. I will clean up and force push.

sleweke-bayer commented:
@hannahlanzrath has put some test cases on sciebo(?). If those work correctly, we can rebase and merge.
However, adding some unit tests before merging would be great.

schmoelder commented:
We need to make sure to also implement the fix for dynamic flow rates provided in #101.

hannahlanzrath commented:
After some name changes of variables to match the MCT model characteristics, the documentation has to be updated. 

- [ ] Remove everything related to film diffusion
- [ ] Change variable names / delete unneeded variables
- [ ] Add cross-section scheme to show the differences to original model with radial zones

I am not entirely sure about how the variable names in cadet are called and could use some clarification. In src/libcadet/model/MultiChannelTransportModel.hpp lines 264-265 

```
		virtual unsigned int numAxialCells() const CADET_NOEXCEPT { return _disc.nCol; }
		virtual unsigned int numRadialCells() const CADET_NOEXCEPT { return _disc.nRad; }
```

was updated to

```
		virtual unsigned int numPrimaryCoordinates() const CADET_NOEXCEPT { return _disc.nCol; }
		virtual unsigned int numSecondaryCoordinates() const CADET_NOEXCEPT { return _disc.nRad; }
```

But since in the documentation the variables are called ncol / nrad, and the return _disc.nCol;/ return _disc.nRad; is it still called with the same name from the python interface using ncol / nrad in the dictionary? 

I think the names in the documentation should match the new primary / secondary coordinates nomenclature or something along these lines, just not sure if I can simply change it or if more code changes are needed.

hannahlanzrath commented:
**TODO:** 

In MultiChannelConvectionDispersionOperator.cpp:

**Change Parameters Names to match Documentation**

- [ ] Change NRAD to NCHANNEL


- [ ] Change RADIAL_DISC_TYPE to CHANNEL_TYPE
keep options 'Equivolume' & 'User_defined'!
- [ ] Remove option 'Equidistant' from CHANNEL_TYPE (former RADIAL_DISC_TYPE) and set 'Equivolume' as default

https://github.com/modsim/CADET/blob/8ef7043dcaa6948a745b117235bd4d42228e37ae/src/libcadet/model/parts/MultiChannelConvectionDispersionOperator.cpp#L757

**Change the way the areas of the Channels is defined**

- [ ] Change RADIAL_COMPARTMENTS to CHANNEL_AREAS / CHANNEL_AREA_FRACTION depending on the method we choose to define the cross section area of each channel:

https://github.com/modsim/CADET/blob/8ef7043dcaa6948a745b117235bd4d42228e37ae/src/libcadet/model/parts/MultiChannelConvectionDispersionOperator.cpp#L765

DECIDE

Proposal 1: CHANNEL_AREAS is a vector of length NCHANNEL filled with scalars that define the absolute value of the cross section area for each channel
Proposal 2: CHANNEL_AREA_FRACTION is a vector of length NCHANNEL filled with fractions (summed up norm to 1) and define the cross section area of each channel as a fraction of the total cross section area of all channels summed up (defined by parameter COL_RADIUS, closer to the current method?)

Reference for COL_RADIUS:

https://github.com/modsim/CADET/blob/8ef7043dcaa6948a745b117235bd4d42228e37ae/src/libcadet/model/parts/MultiChannelConvectionDispersionOperator.cpp#L746

- [ ] Check that the Mass Balance between Channels with differing Cross Section Areas is correct, compare to 2DGRM, if needed ADD Factor that weighs in Channel Volumina/Area Differences

https://github.com/modsim/CADET/blob/8ef7043dcaa6948a745b117235bd4d42228e37ae/src/libcadet/model/parts/MultiChannelConvectionDispersionOperator.cpp#L1051

**Other**

- [] Remove COL_POROSITIES from Code, as it is not needed nor used in the MCT 

https://github.com/modsim/CADET/blob/8ef7043dcaa6948a745b117235bd4d42228e37ae/src/libcadet/model/parts/MultiChannelConvectionDispersionOperator.cpp#L747

@hannahlanzrath 
- [ ] Provide another Jupyter Notebook with test cases Including mass balance test for exchange between channels with differing volumes

hannahlanzrath commented:
**Change the default definition of the cross-section areas**

By default the cross-section areas should be equal and defined by an overall total cross section area.
The vector with the seperate cross section area should remain an optional choice.

My ideas, please feel free to deviate:

- [ ]  This would probably mean going back to the original CHANNEL_AREA_DIST_TYPE (or something along the lines) idea with the option 'EQUIAREA' (Leading to the Use of TOTAL_CHANNEL_AREA) or 'USER_DEFINED' (leading to the use of (CHANNEL_AREAS)

- [ ]  The addition of a parameter CHANNEL_AREA (or TOTAL_CHANNEL_AREA)

hannahlanzrath commented:
Hey guys, @schmoelder @jbreue16 

I have a question regarding the parameters ``GS_TYPE`` , ``MAX_KRYLOV``, ``MAX_RESTARTS`` and ``SCHUR_SAFETY``. Are these still relevant for the MCT? As cadet does not drop an error when I assign nonsensical values such as strings to those parameters  and Eric was not sure how they would come into play in the MCT, the question arose if they can be dropped from the documentation. 
Thanks!

jbreue16 commented:
We don't need those parameters. They're required for the block decomposed linear solver which is only used for models with pores.

 "Is the radial cell the same as a radial zone and is it now analog for the channels, so I can just change radial cell to channel?" - Yes. Maybe we should also rename ``SENS_PARTYPE`` accordingly to something like ``SENS_CHANNEL`` ?

lieres commented:
What would  ``SENS_CHANNEL`` mean? There are no particles in the MCT model.

hannahlanzrath commented:
Since ``SENS_PARTYPE`` is defined in the /input/sensitivity group (https://cadet.github.io/master/interface/sensitivities.html#group-input-sensitivity) I think it should be left as it is but maybe omitted from the MCT documentation, when we are sure it is not needed. My problem is that I do not really understand what it does exactly, but I also think we might not need it in the MCT model?

lieres commented:
I don't think it makes sense to leave it in the code when no particles exist in the model.

jbreue16 commented:
``SENS_PARTYPE`` is (re)used for parameter sensitivities of parameters that depend on the particle type or the radial position (i.e. channel for the MCT), so this should definitely stay in the code. I was just wondering if we should rename the parameter as we have a model with radial position/channels but without particles types. I am not sure what's less confusing.

Iam not an expert with git, maybe we should wait for Johannes. Or, if you want, we can have a look at this tomorrow, shouldn't be too difficult.

hannahlanzrath commented:
I managed to fix my git mistake.

schmoelder commented:
Hi Hannah,
what's the current status on this PR? Please give us a sign when this is ready to merge.
Best
Johannes

jbreue16 commented:
Hi @hannahlanzrath ,

I built the current MCT branch and your notebooks run fine with it.

hannahlanzrath commented:
Then I think we're good to go!

hannahlanzrath commented:
Hello everyone, just because everyone is off and on holiday, before I am gone for a few weeks I wanted to ask what is the current status and plan with merging the MCT?

schmoelder commented:
Hi Hannah,

sorry for the long silence. 

If we look at the to-dos, we're only missing a unit test. Maybe we can get this done next Friday during our Core-Dev Meeting. Other than that, we're good to go.

hannahlanzrath commented:
[MCT_demo.zip](https://github.com/modsim/CADET/files/12380221/MCT_demo.zip)

hannahlanzrath commented:
![image](https://github.com/modsim/CADET/assets/86779218/2a272d6e-eac2-44ff-8400-6d48aaaa7ec5)

hannahlanzrath commented:
**Update on the Unittest Issue**

A Unittest file (test->MultiChannelTransportModel.cpp) was added, and a simple test case created, that compares the outlets of two seperate channels with the same configuration. 

- [ ] From there on, more test cases can be added (I think I could do that, maybe with a little help of someone knows the language better)
- [ ] The test cases from the Jacobian must be modified so they use the MCT file (? I am not sure how that would be solved)

Also we rebased as a preparation to merge into the main branch

hannahlanzrath commented:
What would be the right way to pull now to avoid merge conflicts and just get the current version?

hannahlanzrath commented:
Today Jan and I worked on the Unittests:

- We added a Unittest that Compares the Output of two simulations Running Exchange from Channel 1 to Channel 2 and vice versa
- In JasonTestModels.cpp we added the case MULTI_CHANNEL_TRANSPORT with NCHANNEL and CHANNEL_CROSS_SECTION_AREAS for the Jacobian Tests and added them in test->MultiChannelTransportModel.cpp of which some work and some don't
- We added the MCT as well for the LWE specifying NCHANNEL=1 
- In test->MultiChannelTransportModel.cpp we added a Unittest for backwards flow based on the one for the other models, butit fails -> We tested forwards and backward flow and looked at the output graphically, for the backward flow it shows artifacts in the beginning even with high ncol and abs/rel tolerances for the solver. Ideas?
-
![image](https://github.com/modsim/CADET/assets/86779218/8bd4f262-ecb8-42a9-9918-83eb62fb6550)

![image](https://github.com/modsim/CADET/assets/86779218/e1e6a19a-07f2-4690-82a3-5c841db760a6)

TODO:

- Correct Port connections in MultiChannelTranportModel.cpp to be applicable for each test case

hannahlanzrath commented:
Jan and I found out that the Jacobian tests fail for the GRM2D as well. The fowards/backwards tests do not work for any of the models, we even tested on a branch without MCT. Thus the tests not working likely has nothing to do with the implementation of the MCT but a general problem of the cadet-core. We will check the commits if we find a state in which these tests were still working.

hannahlanzrath commented:
Sam added, that for the 2DGRM Algorithmic Differentiation is not working, so

- [ ] Comment out AD tests for MCT and 2DGRM

hannahlanzrath commented:
Johannes and I noticed, that the Exchange Matrix is not only nchannel*nchannel but can be defined for every component, so

- [ ] Exchange Matrix Documentation in Interface should say size nchannel*nchannel*ncomp

jbreue16 commented:
Ive fixed the MCT backwards flow and updated the tests:
We now have tests for simple logics: fwd vs bwd flow; fwd vs bwd exchange; two channels with same inlet yield same result.

We should add a test against a reference solution. @hannahlanzrath I think its alright to use the LRM without porosity to generate one. Maybe @sleweke can generate an analytical one using CADET-semianalytic sometime.

ronald-jaepel commented:
Heads up: 

Before merging this branch would have to be rebased anyways, so we went ahead an rebased it now to get it to play nicely with the update build instructions. This will, however, break workflows for people using this branch with the old build instructions (I think). Therefore I haven't force pushed the rebased branch onto this branch but create a new one. Let's discuss in which order to merge these changes Wednesday in the office.

schmoelder commented:
Congrats! :tada:

hannahlanzrath commented:
thanks a lot for all your work on this guys, you're the coolest people 😎👌
